### PR TITLE
V2.16

### PR DESCRIPTION
**Pull request by KyleDavisSA**
This PR updates the master branch to be compatible with the latest CalculiX v2.16.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -4,7 +4,7 @@
 
 The adapter was initially developed for conjugate heat transfer (CHT) simulations via preCICE by Lucia Cheung in the scope of her master’s thesis [[1]](https://www5.in.tum.de/pub/Cheung2016_Thesis.pdf) in cooperation with [SimScale](https://www.simscale.com/). For running the adapter for CHT simulations refer to this thesis. The adapter was extended to fluid-structure interaction by Alexander Rusch [[2]](https://www.gacm2017.uni-stuttgart.de/registration/Upload/ExtendedAbstracts/ExtendedAbstract_0138.pdf).
 
-This adapter was developed for **CalculiX version 2.15.** Other versions may be compatible, yet they have not been tested. Please let us know if you want to use a different version.
+This adapter was developed for **CalculiX version 2.16.** Other versions may be compatible, yet they have not been tested. Please let us know if you want to use a different version.
```
Comment (uekerman) [inline README.md]: We could explain here that other versions are available in other branches.

File: adapter/CCXHelpers.c [hunk 2]
```diff
@@ -80,7 +83,10 @@ void getNodeCoordinates( ITG * nodes, ITG numNodes, int dim, double * co, double
 	{
 		int nodeIdx = nodes[i] - 1;
 		//The displacements are added to the coordinates such that in case of a simulation restart the displaced coordinates are used for initializing the coupling interface instead of the initial coordinates
-		for( j = 0 ; j < dim ; j++ ) coordinates[i * dim + j] = co[nodeIdx * 3 + j] + v[nodeIdx * mt + j + 1];
+		for( j = 0 ; j < dim ; j++ )
+    {
+      coordinates[i * dim + j] = co[nodeIdx * 3 + j] + v[nodeIdx * mt + j + 1];
+    }
```
Comment (uekerman) [inline adapter/CCXHelpers.c]: Formatting looks wrong here.

File: README.md (hunk 1 above)
Comment (uekerman) [reply README.md]: And where do we explain which preCICE version is compatible?

File: adapter/CCXHelpers.c (hunk 2 above)
Comment (nkr0) [reply adapter/CCXHelpers.c]: we could run precice's clang-format conf file on everything inside adapter/.

**Comment by KyleDavisSA** [comment]
@MakisH The tutorial works for me. It doesn't look like a Calculix issue there. Did you try other mapping? Is it only that tutorial that does not work?

File: adapter/CCXHelpers.c [hunk 3]
```diff
@@ -80,7 +83,10 @@ void getNodeCoordinates( ITG * nodes, ITG numNodes, int dim, double * co, double
 	{
 		int nodeIdx = nodes[i] - 1;
 		//The displacements are added to the coordinates such that in case of a simulation restart the displaced coordinates are used for initializing the coupling interface instead of the initial coordinates
-		for( j = 0 ; j < dim ; j++ ) coordinates[i * dim + j] = co[nodeIdx * 3 + j] + v[nodeIdx * mt + j + 1];
+		for( j = 0 ; j < dim ; j++ )
+    		{
+      			coordinates[i * dim + j] = co[nodeIdx * 3 + j] + v[nodeIdx * mt + j + 1];
```
Comment (MakisH) [inline adapter/CCXHelpers.c:88]: We could fix the indentation in this file

**Comment by MakisH** [comment]
@KyleDavisSA I tried again, I still get errors, only with RBF (NN and NP work as expected). I am using preCICE v2.0.2, OpenFOAM v1912, and PETSc  3.9.1. I also reproduced this in the 3DTube tutorial (switching to rbf-thin-plate-splines). On the Fluid side (which does both mappings), I get a similar error:

<details>
<summary>Click to reveal...</summary>
<p>


```
---[preciceAdapter] [DEBUG] Writing coupling data...
---[preciceAdapter] [DEBUG] Advancing preCICE...
---[precice]  Mapping Forces0 conservative from Fluid-Mesh-Faces (ID 1) to Calculix_Mesh (ID 2) for dimension 0) with polynomial set to separate
---[precice]  Mapping Forces0 conservative from Fluid-Mesh-Faces (ID 1) to Calculix_Mesh (ID 2) for dimension 1) with polynomial set to separate
---[precice]  Mapping Forces0 conservative from Fluid-Mesh-Faces (ID 1) to Calculix_Mesh (ID 2) for dimension 2) with polynomial set to separate
---[precice]  Compute read mapping from mesh "Calculix_Mesh" to mesh "Fluid-Mesh-Nodes".
---[precice]  Using tree-based preallocation for matrix C
---[precice]  Using tree-based preallocation for matrix A
---[precice]  Mapping DisplacementDeltas0 consistent from Calculix_Mesh (ID 2) to Fluid-Mesh-Nodes (ID 0) for dimension 0) with polynomial set to separate
KSP Object: Coefficient Solver (solverC_) 1 MPI processes
  type: gmres
    restart=30, using Classical (unmodified) Gram-Schmidt Orthogonalization with no iterative refinement
    happy breakdown tolerance 1e-30
  maximum iterations=10000, nonzero initial guess
  tolerances:  relative=1e-09, absolute=1e-50, divergence=10000.
  left preconditioning
  using PRECONDITIONED norm type for convergence test
PC Object: (solverC_) 1 MPI processes
  type: icc
    out-of-place factorization
    0 levels of fill
    tolerance for zero pivot 2.22045e-14
    using Manteuffel shift [POSITIVE_DEFINITE]
    matrix ordering: natural
    factor fill ratio given 1., needed 1.
      Factored matrix follows:
        Mat Object: 1 MPI processes
          type: seqsbaij
          rows=2588, cols=2588
          package used to perform factorization: petsc
          total: nonzeros=3350166, allocated nonzeros=3350166
          total number of mallocs used during MatSetValues calls =0
              block size is 1
  linear system matrix = precond matrix:
  Mat Object: C 1 MPI processes
    type: seqsbaij
    rows=2588, cols=2588
    total: nonzeros=3350166, allocated nonzeros=3350166
    total number of mallocs used during MatSetValues calls =0
        block size is 1
---[precice] ERROR:  RBF linear system has not converged. Try to fix axis-aligned mapping setups by marking perpendicular axis as dead.
```
</p>
</details>

The preCICE tests pass. My runs are in serial.

I attach here the log files (with traces) for the flap_perp tutorial:
[Fluid.log](https://github.com/precice/calculix-adapter/files/4443615/Fluid.log)
[Solid.log](https://github.com/precice/calculix-adapter/files/4443616/Solid.log)

**Comment by MakisH** [comment]
I can also use RBF mapping in the OpenFOAM-OpenFOAM flow-over-plate tutorial without any problems.

I don't have any issues with the current master (73a9a062e3cd3a0c8d01d480950fc7f3172037a4) or develop (8e954db79e68feb47792de2bd8e0aec610a428af) - CalculiX 2.15.

**Comment by MakisH** [comment]
@KyleDavisSA I tried again and I am still getting the same error, also in a different system. This time I tried on a clean install on Ubuntu 20.04, preCICE 2.0.1 from master built in Debug mode, PETSc 3.12 from APT (3.12.4+dfsg1-1), CalculiX 2.16 and the adapter from the `v2.16` branch.

Running the `tutorials/FSI/flap_perp/OpenFOAM-CalculiX` tutorial without any changes, with OpenFOAM v1912, returns on the Fluid side:

```
---[preciceAdapter] [DEBUG] Writing coupling data...
---[preciceAdapter] [DEBUG] Advancing preCICE...
---[precice]  Mapping Forces0 conservative from Fluid-Mesh-Faces (ID 0) to Solid (ID 2) for dimension 0) with polynomial set to separate
---[precice]  Mapping Forces0 conservative from Fluid-Mesh-Faces (ID 0) to Solid (ID 2) for dimension 1) with polynomial set to separate
---[precice]  Mapping Forces0 conservative from Fluid-Mesh-Faces (ID 0) to Solid (ID 2) for dimension 2) with polynomial set to separate
---[precice]  Compute read mapping from mesh "Solid" to mesh "Fluid-Mesh-Nodes".
---[precice]  Using tree-based preallocation for matrix C
---[precice]  Using tree-based preallocation for matrix A
---[precice]  Mapping Displacements0 consistent from Solid (ID 2) to Fluid-Mesh-Nodes (ID 1) for dimension 0) with polynomial set to separate
KSP Object: QR Solver 1 MPI processes
  type: lsqr
    standard error not computed
    using inexact matrix norm
  maximum iterations=10000, initial guess is zero
  tolerances:  relative=1e-05, absolute=1e-50, divergence=10000.
  left preconditioning
  using UNPRECONDITIONED norm type for convergence test
PC Object: 1 MPI processes
  type: none
  linear system matrix = precond matrix:
  Mat Object: Q 1 MPI processes
    type: seqdense
    rows=86, cols=4
    total: nonzeros=344, allocated nonzeros=344
    total number of mallocs used during MatSetValues calls=0
---[precice] ERROR:  Polynomial QR linear system has not converged. Try to fix axis-aligned mapping setups by marking perpendicular axis as dead.
```

Again, running the same exact tutorial on the v2.15 is perfectly fine.

Note that, for whatever reason, [Travis did not check this case](https://travis-ci.org/github/precice/systemtests/builds/693127009). @Eder-K any clue why?

**Comment by Eder-K** [comment]
Good catch, I think this happens because we [do not list the test in the trigger_systemtests script](https://github.com/precice/systemtests/blob/e03f675945ed6ef3c7aa28e7bb0df463873fd713/trigger_systemtests.py#L22). I will open a PR and check if these run on Travis.

**Pull request merged**